test/widgets/test_moc.py

```diff
@@ -0,0 +1,193 @@
+
+
+def no_op(*args, **kwargs):
+    pass
+
+
+class FakeWindow:
+    class _NestedWindow:
+        wid = 10
+    window = _NestedWindow()
+
+
+@pytest.fixture
+def patched_moc(fake_qtile, monkeypatch):
+    widget = moc.Moc()
+    MockMocpProcess.reset()
+    monkeypatch.setattr(widget, "call_process", MockMocpProcess.run)
+    monkeypatch.setattr("libqtile.widget.moc.subprocess.Popen", MockMocpProcess.run)
+    fakebar = Bar([widget], 24)
+    fakebar.window = FakeWindow()
+    fakebar.width = 10
+    fakebar.height = 10
+    fakebar.draw = no_op
+    widget._configure(fake_qtile, fakebar)
+    return widget
+
+
+def test_moc_poll_string_formatting(patched_moc):
+
+    # Both artist and song title
+    assert patched_moc.poll() == "♫ Rick Astley - Never Gonna Give You Up"
+
+    # No artist
+    MockMocpProcess.info[0]["Artist"] = ""
+    assert patched_moc.poll() == "♫ Never Gonna Give You Up"
+
+    # No title
+    MockMocpProcess.info[0]["SongTitle"] = ""
+    assert patched_moc.poll() == "♫ rickroll"
+
+
+def test_moc_state_and_colours(patched_moc):
+
+    # Initial poll - playing
+    patched_moc.poll()
+    assert patched_moc.layout.colour == patched_moc.play_color
+
+    # Toggle pause
+    patched_moc.play()
+    patched_moc.poll()
+    assert patched_moc.layout.colour == patched_moc.noplay_color
+
+    # Toggle pause --> playing again
+    patched_moc.play()
+    patched_moc.poll()
+    assert patched_moc.layout.colour == patched_moc.play_color
+
+
+def test_moc_button_presses(manager_nospawn, minimal_conf_noscreen, monkeypatch):
+
+    # This needs to be patched before initialising the widgets as mouse callbacks
+    # bind subprocess.Popen.
+    monkeypatch.setattr("subprocess.Popen", MockMocpProcess.run)
+
+    # Long interval as we don't need this polling on its own.
+    mocwidget = moc.Moc(update_interval=30)
+    MockMocpProcess.reset()
+    monkeypatch.setattr(mocwidget, "call_process", MockMocpProcess.run)
+    monkeypatch.setattr("libqtile.widget.moc.subprocess.Popen", MockMocpProcess.run)
+
+    config = minimal_conf_noscreen
+    config.screens = [
+        libqtile.config.Screen(
+            top=libqtile.bar.Bar([mocwidget], 10)
+        )
+    ]
+
+    manager_nospawn.start(config)
+
+    # When started, we have the first item playing
+    topbar = manager_nospawn.c.bar["top"]
+    assert topbar.info()["widgets"][0]["text"] == "♫ Rick Astley - Never Gonna Give You Up"
+
+    # Trigger next item and wait for update poll
+    topbar.fake_button_press(0, "top", 0, 0, button=4)
+    _, result = manager_nospawn.c.widget["moc"].eval("self.poll()")
+    assert result == "♫ Neil Diamond - Sweet Caroline"
+
+    # Trigger next item and wait for update poll
+    # This item's state is set to "STOP" so there's no track title
+    topbar.fake_button_press(0, "top", 0, 0, button=4)
+    _, result = manager_nospawn.c.widget["moc"].eval("self.poll()")
+    assert result == "♫"
+    # assert topbar.info()["widgets"][0]["text"] == "♫"
```

Comment:
Should we uncomment this?

Reply:
No, we should delete it. `poll` returns the text value, rather than updating the display. So, we can check the text by looking at the return value but we can't check the `text` value. 

Well, that's not quite true, I could ask the manager to eval `self.update(self.poll())`... Maybe that's better.